src/v/pandaproxy/schema_registry/schema_id_cache.h

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "config/property.h"
+#include "model/fundamental.h"
+#include "pandaproxy/schema_registry/subject_name_strategy.h"
+#include "pandaproxy/schema_registry/types.h"
+#include "seastarx.h"
+
+#include <boost/multi_index/indexed_by.hpp>
+#include <boost/multi_index/key.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/sequenced_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+#include <iterator>
+#include <utility>
+
+namespace pandaproxy::schema_registry {
+
+///\brief An MRU cache of valid schemas for a given topic.
+class schema_id_cache {
+public:
+    enum class field : uint8_t { key, val };
+    using offsets_t = std::optional<std::vector<int32_t>>;
+
+    explicit schema_id_cache(config::binding<size_t> cap)
+      : _capacity{std::move(cap)} {
+        _capacity.watch([this]() { shrink_to_capacity(); });
+    }
+
+    bool has(
+      model::topic_view topic,
+      field field,
+      subject_name_strategy sns,
+      schema_id s_id,
+      offsets_t const& offsets) {
+        auto& map = _cache.get<underlying_map>();
+        auto it = map.find(entry::view_t(topic, field, sns, s_id, offsets));
+        bool has = it != map.end();
+        if (has) {
+            auto& list = _cache.get<underlying_list>();
+            list.relocate(list.begin(), _cache.project<underlying_list>(it));
```

Comment:
> A little unclear from boost's documentation. Does this just 'shift' everything? e.g.

Yes, it will move it to the front of the "list", and seat it appropriately in the "set" by reevaluating the key, which hasn't changed.